### ISSUE TITLE
DualShock 4 Gyro not working on Bluetooth

### ISSUE THREAD

**Issue by EverSalazar** [JibbSmart/JoyShockMapper#54]
Hi, I know you mentioned that Bluetooth support for DS4 is new and might not work and that's exactly my case. It works perfectly via USB, but the gyro doesn't respond over Bluetooth. Are there any known steps to troubleshoot this?

**Specs:**
OS: Windows 10 Home version 1909
Processor: AMD Ryzen 5 2600
Motherboard: MSI AMD X470
Bluetooth adapter: Kensington USB-Bluetooth Micro 4.0 connected to a USB 3.0 port
Control: Dualshock 4 model CUH-ZCT2U

And thanks for JoyShockMapper. It's beautiful the work you've done.

**Comment by JibbSmart** [comment]
Thanks for letting me know. I'm working on a fix. I haven't been able to figure out the exact cause of the problem, but I think I might have the solution anyway.

**Comment by ottoaho** [comment]
I am unable to get DS4 gyro inputs with Bluetooth or with USB. All buttons and sticks register just fine with both connection methods. I tried disconnecting my Logitech K400 Plus keyboard+mouse (disconnected its bluetooth dongle) in case it was interfering but that didn't help.

**Comment by JibbSmart** [comment]
Hi @ottoaho, this is very unusual. Usually if it doesn't work on USB it's actually that the user hasn't enabled gyro controls, so just in case this is the case here: While connected by USB, have you tried setting GYRO_SENS = 1 and seeing if the mouse moves as you move the controller?

**Comment by ottoaho** [comment]
> Hi @ottoaho, this is very unusual. Usually if it doesn't work on USB it's actually that the user hasn't enabled gyro controls, so just in case this is the case here: While connected by USB, have you tried setting GYRO_SENS = 1 and seeing if the mouse moves as you move the controller?

You are right, I had overlooked this setting. Would it be a good idea to have GYRO_SENS set to 1 in the initial config that gets loaded at program startup?

Thanks @JibbSmart for the reply and for the wonderful piece of software.

**Comment by ShogunPatch** [comment]
Hi JibbSmart
First of all, congrats on what looks like a fantastic project, which I came across after watching a Youtube Video by Linus Tech Tips.

I am not a programmer, or even a very good gamer (sadly), but I do enjoy playing Fortnite on PC with a controller with my son.  My aiming, however, is atrocious so the idea of being able to use the gyros to aim is very interesting.

I have successfully installed and run JSM but, like the posters above, have found that the gyros only seem to pickup if connected via USB, not via Bluetooth.  I can see you are aware of and working on that so look forward to seeing a fix.  However, I have a secondary issue which is that even when I am connected via USB and can see in the JSM console that my gyro movements are being picked up, they don't seem to do anything on Fortnite.  Any ideas?

I have tried both of the Fortnite configs in the Wiki to no avail.

FYI, I typically play using GeForceNow:  thought that might be an issue so have also tested this playing natively on my machine but still no go.

**Specs:**
Toshiba Z30
Intel i5-4200U @ 1600GHz 2.30GHz
16.0GB Ram

Windows 10 Pro 19041.572

Playing with a (new) DS4 v2

**Comment by dwang7** [comment]
I was having issues getting ds4 gyro to work under bluetooth as well.  What got it working for me was to enable ds4 controller support under steam.  Once that was configured, joyshockmapper was able to recognize the gyro readings when ds4 is connected via bluetooth.

**Comment by ShogunPatch** [comment]
> I was having issues getting ds4 gyro to work under bluetooth as well. What got it working for me was to enable ds4 controller support under steam. Once that was configured, joyshockmapper was able to recognize the gyro readings when ds4 is connected via bluetooth.

Hi dwang7, I've only just seen this but I'm so glad I have:  I've just installed and setup Steam and it seems to have fixed it for me, too, which is great!  Thank you so much.  

Jibb, I now have the gyro functionality working over Bluetooth but now I am struggling to get to grips with all of the added functionality & configs:  if I set it up to use either of the Fortnite configs from the gyrowiki, I am all over the place.  As a pretty low-level Fortnite player (which is all I use this for) all I really want in the first instance is for my controller to behave as a normal DS4 most of the time, with the one exception that when I press the aim button (ZL) it temporarily toggles on the gyro so I can use gyro/motion control to (hopefully) improve my aim, toggling back off again when I release the aim button.  I realise such a setup is really basic and would only make use of a fraction of the power of Joyshockmapper, but TBH at the moment I am a bit overwhelmed with it all and am struggling to use it at all as a result.  
I'm not a coder, however, and whilst I assume this should be really easy to set up, I can't for the life of me get it to work.  I've tried entering the commend ZL = GYRO_OFF but it doesn't actually seem to do anything.  Could you please explain to me like I'm a five-year-old how to set this up?
Many thanks.

**Comment by ShogunPatch** [comment]
Actually, scratch that - a bit more fiddling and trial and error on my part and I have now managed to get it to work! Brilliant!
I have to say, having previously tried various solutions to try and get gyro aiming to work for my particular setup, including DS4Windows and others and been reduced to tearing my hair out on each occasion, this is the first and only one that has worked at all playably, and not only playably but actually seems to work well.  
JSM may not have a sexy GUI interface, and may have a reasonably steep learning curve for non-coders like me, but I am really pleased I have managed to get it to work, and really impressed, Jibb, with what you have produced.  Keep up the good work.
